### PR TITLE
Add support for PEP 705

### PR DESCRIPTION
**Pull request by JelleZijlstra**

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Let's not merge this until python/peps#3440 is merged.

File: doc/index.rst
Comment (srittau) [inline doc/index.rst:410]: I don't understand this sentence. So,

```python
class Super(TypedDict, other_keys=False): ...
class Sub(TypedDict, other_keys=True): ...
```

Does that mean that `Sub` semantically(?) doesn't allow other keys (but practically it does?) and `Sub.__other_keys__` is `False`?

Comment (JelleZijlstra) [reply doc/index.rst:410]: Your understanding is correct, but I agree this would be confusing. I'll leave the decision to @alicederyn as the PEP author, but I'd recommend one of these two:

- Make it an error for an `other_keys=True` TypedDict to inherit from an `other_keys=False` TypedDict. This would mean that any subclass of an `other_keys=False` TypedDict must also set `other_keys=False`. This is my preference as it's most explicit.
- Automatically set `other_keys=False` if any base class has `other_keys=False`.

File: doc/index.rst
Comment (alicederyn) [inline doc/index.rst:323]: I've been trying to stick to saying that the "entry" is readonly, rather than "key", to avoid possible confusion.

File: doc/index.rst
Comment (alicederyn) [reply doc/index.rst:410]: Good point. I've also gone for explicit, and updated PEP-705.

File: src/typing_extensions.py
Comment (AlexWaygood) [inline src/typing_extensions.py:883]:
```suggestion
                        f"TypedDict {base.__name__!r}"
```

(we know that `base` is an instance of `type` at this point in the function, so the `.__name__` attribute access should be safe)

File: src/typing_extensions.py
Comment (AlexWaygood) [inline src/typing_extensions.py:888]: Very minor, but a `set` might be better here (it would make the `if Required in qualifiers` check below O(1) )

```suggestion
                qualifiers = set(_get_typeddict_qualifiers(annotation_type))
```

File: src/typing_extensions.py
Comment (AlexWaygood) [inline src/typing_extensions.py:869]: runs about twice as fast:

```suggestion
                        own_annotations.keys() -
                        base.__dict__.get('__annotations__', {}).keys()
```

File: doc/index.rst
Comment (alicederyn) [reply doc/index.rst:323]: The PEP got rewritten to use "item" instead of "entry", as that's what the Python docs/API use elsewhere.

File: src/typing_extensions.py
Comment (alicederyn) [inline src/typing_extensions.py:912]: Do you need to remove the key from readonly_keys too here? It might be inherited.

**Comment by alicederyn** [comment]
PEP-705 has dropped both `readonly` and `other_keys`; only the `typing.ReadOnly` annotation needs to be added now.

File: doc/index.rst
Comment (JelleZijlstra) [reply doc/index.rst:410]: No longer relevant as `other_keys` was removed

File: src/typing_extensions.py
Comment (JelleZijlstra) [reply src/typing_extensions.py:884]: This code is gone

File: src/typing_extensions.py
Comment (JelleZijlstra) [reply src/typing_extensions.py:889]: Done

File: src/typing_extensions.py
Comment (JelleZijlstra) [reply src/typing_extensions.py:870]: Removed this code

File: src/typing_extensions.py
Comment (JelleZijlstra) [reply src/typing_extensions.py:913]: This is actually already buggy for Required/NotRequired, even in CPython:

```
Python 3.13.0a2+ (heads/pep702:6ab4584ac4, Nov 26 2023, 17:26:43) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import *
>>> class A(TypedDict):
...     a: NotRequired[int]
... 
>>> class B(A):
...     a: Required[int]
... 
>>> B.__required_keys__
frozenset({'a'})
>>> B.__optional_keys__
frozenset({'a'})
```

I'll leave fixing that to a separate PR.

Comment (JelleZijlstra) [reply src/typing_extensions.py:913]: (But I'll fix this for ReadOnly.)

Comment (alicederyn) [reply src/typing_extensions.py:886]: I think that example is illegal though. You can't change the required/notrequired on a key like that; it's banned by PEP-589. It's only once we have ReadOnly that this becomes an issue.

File: src/typing_extensions.py
Comment (alicederyn) [inline src/typing_extensions.py:1980]:
```suggestion
        There is no runtime checking for this property.
```

File: src/typing_extensions.py
Comment (alicederyn) [inline src/typing_extensions.py:1969]:
```suggestion
        """A special typing construct to mark an item of a TypedDict as read-only.
```
If we want to keep to the naming approach of the PEP

File: src/typing_extensions.py
Comment (JelleZijlstra) [reply src/typing_extensions.py:886]: Replied about this on CPython

File: src/typing_extensions.py
Comment (AlexWaygood) [inline src/typing_extensions.py:1970]:
```suggestion

        For example:
```

File: src/typing_extensions.py
Comment (AlexWaygood) [inline src/typing_extensions.py:1995]:
```suggestion

        For example:
```

**Review by AlexWaygood — APPROVED** [review]
Some minor suggestions (mostly optional), but overall this looks good!

**Review by AlexWaygood — APPROVED** [review]
LGTM

**Pull request merged**
